### PR TITLE
[WIP] Port provider to new LXD client

### LINKED ISSUE #70: Port to the new LXD client package
Hey there,

We've recently implemented a new LXD client package an are phasing out our old one (client.go).
LXD itself has now been entirely ported onto the new client code and we're planning on completely removing the old one from the tree with LXD 2.16.

The new client package comes with what should be a future proof API design where feature additions won't cause existing users to break and need modifications.

You can find the Go documentation for it here:  http://godoc.org/github.com/lxc/lxd/client

I noticed you're also parsing the LXD command line tool's configuration to get information on the remotes.
To do so with the new package, you should be using: http://godoc.org/github.com/lxc/lxd/lxc/config

Let me know if you have any questions or need help porting. I've ported all of our own tools and it should be reasonably painless. https://github.com/lxc/lxd/pull/3422 should be full of examples on how to do it.

### PR DESCRIPTION
**Pull request by sl1pm4t**
Fixes #70

### CODE REVIEW

**Comment by jtopjian** [comment]
There seems to be a lot of random failures going on. At first glance, it feels like there's a timing issue somewhere: sometimes a container is successfully destroyed, other times it hangs indefinitely.

**Comment by sl1pm4t** [comment]
Yeah, I'm wondering if it could be that I'm using `op.Wait()` now in place of the old polling refresh. If the `op.Wait()` are returning sooner than a poll cycle would have detected a new state, then maybe it's hitting create vs destroy race conditions not previously seen.

**Comment by jtopjian** [comment]
I've been playing around with this using that `deploy.sh` script I just submitted. I haven't had a case of a test failing to finish, but there are still failures. Fortunately the failing tests are consistent between runs.

Curiously, when I was testing on AWS, I was seeing the same 3 tests fail. Now on my OpenStack environment, it's only a "file" test. It could be that I'm pulling the LXD server package from the git-master ppa... but that would be one crazy coincidence of timing :)

**Comment by jtopjian** [comment]
Yes! It's a timing issue.

I added `time.Sleep(10 * time.Second)` between when `op2.Wait()` returns and before files are uploaded and the file was successfully uploaded.

I'm sure this isn't specific to just file uploads.

**Comment by sl1pm4t** [comment]
Ok cool - so it's mostly a synthetic problem introduced by the tests.

**Comment by jtopjian** [comment]
I wouldn't say it was strictly introduced by tests. The same file upload failures happened when I was testing a standalone `main.tf` file with a compiled binary.

**Comment by jtopjian** [comment]
Continuing my liveblogging, I added `time.Sleep(1 * time.Second)` to a few places where the container transitions state and all tests passed. Going to open a PR and see if Travis likes it.

**Comment by jtopjian** [comment]
It's a squashfs problem. Enough hammering at the Directory backend and the system will get wedged (as well as do other weird things like not upload files and such, which I'm now thinking were just file calls that were lost).

Re-adding the `WaitForState` stuff helps in that it adds some padding between API calls.

I'm hesitant to say this is unique to Travis or tests, though. I'm sure other users will run into this as well. And even though the underlying issue is a storage driver problem, perhaps having Terraform add some artificial padding will help prevent other LXD servers from becoming wedged.

**Comment by sl1pm4t** [comment]
Awesome - how did you figure out it was squashfs related? I've never dug into the innards of squashfs

**Comment by jtopjian** [comment]
> how did you figure out it was squashfs related?

I'd love to claim amazing powers of low-level system debugging, but the dead giveaway was LXD being locked up beyond repair and the process table was full of hung `unsquashfs` processes. I was able to inevitably recreate this on newly created vms just by hammering away at tests or individual configuration files. :)

**Comment by sl1pm4t** [comment]
@jtopjian - before we merge this PR - do you want me to send a PR to your lxdhelpers repo with the changes I had to make?

**Comment by jtopjian** [comment]
I was literally just thinking about that. If I merge those changes, it'll break gobetween as gobetween doesn't vendor dependencies. I'll have to get a PR submitted to that repo first. Let's use your modified version for now and I'll work on gobetween within the next week or so.

File: lxd/provider.go [hunk 1]
```diff
@@ -367,3 +348,66 @@ func validateLxdRemoteScheme(v interface{}, k string) ([]string, []error) {
...
+	if p.clientMap == nil {
+		p.clientMap = make(map[string]lxd.Server)
+	}
+
+	p.clientMap[remote] = client
+	return client, nil
+}
+
+// GetClient returns an LXD client for the named remote
+func (p *LxdProvider) GetClient(remote string) (lxd.ContainerServer, error) {
```
Comment (sl1pm4t) [inline lxd/provider.go]: The naming of this and `GetServerClient()` below are kind of confusing and backwards - I should improve.

File: lxd/provider.go [hunk 2]
```diff
@@ -367,3 +348,66 @@ func validateLxdRemoteScheme(v interface{}, k string) ([]string, []error) {
...
+		return nil, err
+	}
+	ci, err := s.GetConnectionInfo()
+	if ci.Protocol == "simplestreams" || ci.Protocol == "lxd" {
+		return s.(lxd.ImageServer), nil
+	}
+	return nil, fmt.Errorf("remote (%s / %s / %s) is not an ImageServer", remote, ci.Addresses[0], ci.Protocol)
+}
+
+// GetClient returns an LXD client for the named remote
```
Comment (sl1pm4t) [inline lxd/provider.go]: func name is wrong here

File: lxd/resource_lxd_cached_image.go [hunk 3]
```diff
@@ -88,41 +90,57 @@ func resourceLxdCachedImageCreate(d *schema.ResourceData, meta interface{}) erro
...
 
 	// Execute the copy
-	err = srcClient.CopyImage(image, dstClient, copyAliases, aliases, false, false, resourceLxdCachedImageCopyProgressHandler)
+	// Image copy arguments
+	args := lxd.ImageCopyArgs{
+		Aliases: aliases,
+		Public:  false,
+	}
+
+	op, err := dstClient.CopyImage(srcClient, *imgInfo, &args) //dstClient, copyAliases, aliases, false, false, resourceLxdCachedImageCopyProgressHandler)
```
Comment (sl1pm4t) [inline lxd/resource_lxd_cached_image.go]: needs cleanup

File: lxd/resource_lxd_cached_image.go [hunk 4]
```diff
@@ -166,12 +184,24 @@ func resourceLxdCachedImageUpdate(d *schema.ResourceData, meta interface{}) erro
...
+			err := client.DeleteImageAlias(alias)
+			if err != nil {
+				return err
+			}
 		}
 		// Add new
 		for _, a := range aliasesToAdd.List() {
 			alias := a.(string)
-			client.PostAlias(alias, "", id.fingerprint)
+			// client.PostAlias(alias, "", id.fingerprint)
```
Comment (sl1pm4t) [inline lxd/resource_lxd_cached_image.go]: needs cleanup

File: lxd/resource_lxd_container.go [hunk 5]
```diff
@@ -203,40 +210,84 @@ func resourceLxdContainerCreate(d *schema.ResourceData, meta interface{}) error
...
+
+	// Even though op.Wait has completed,
+	// wait until we can see the container is running via a new API call.
+	// At a minimum, this adds some padding between API calls.
 	stateConf := &resource.StateChangeConf{
 		Target:     []string{"Running"},
 		Refresh:    resourceLxdContainerRefresh(client, name),
 		Timeout:    3 * time.Minute,
-		Delay:      refreshInterval,
+		Delay:      3 * time.Second,
```
Comment (sl1pm4t) [inline lxd/resource_lxd_container.go]: Any value in going back to `refreshInterval`?

File: lxd/resource_lxd_container.go [hunk 6]
```diff
@@ -420,36 +474,48 @@ func resourceLxdContainerDelete(d *schema.ResourceData, meta interface{}) (err e
 		return err
 	}
 
-	refreshInterval := p.RefreshInterval
+	// refreshInterval := p.RefreshInterval
```
Comment (sl1pm4t) [inline lxd/resource_lxd_container.go]: tidy up

File: lxd/resource_lxd_container.go [hunk 7]
```diff
@@ -420,36 +474,48 @@ func resourceLxdContainerDelete(d *schema.ResourceData, meta interface{}) (err e
...
-		// Wait until the container is in a Stopped state
+		// Even though op.Wait has completed,
+		// wait until we can see the container has stopped via a new API call.
+		// At a minimum, this adds some padding between API calls.
 		stateConf := &resource.StateChangeConf{
 			Target:     []string{"Stopped"},
 			Refresh:    resourceLxdContainerRefresh(client, name),
 			Timeout:    3 * time.Minute,
-			Delay:      refreshInterval,
+			Delay:      3 * time.Second,
```
Comment (sl1pm4t) [inline lxd/resource_lxd_container.go]: same as above

File: lxd/resource_lxd_container.go [hunk 8]
```diff
@@ -522,6 +588,7 @@ func resourceLxdContainerUploadFile(client *lxd.Client, container string, file i
 	}
 
 	mode := os.FileMode(0755)
+	// mode := 755
```
Comment (sl1pm4t) [inline lxd/resource_lxd_container.go]: tidy up

File: lxd/resource_lxd_container.go [hunk 9]
```diff
@@ -532,21 +599,30 @@ func resourceLxdContainerUploadFile(client *lxd.Client, container string, file i
 			return fmt.Errorf("Could not determine file mode %s", v)
 		}
 
+		// mode = int(m)
```
Comment (sl1pm4t) [inline lxd/resource_lxd_container.go]: tidy up

File: lxd/resource_lxd_container.go [hunk 10]
```diff
@@ -570,11 +646,60 @@ func resourceLxdContainerDeleteFile(client *lxd.Client, container string, file i
...
+	if err := client.DeleteContainerFile(container, fileTarget); err != nil {
 		return fmt.Errorf("Could not delete file %s: %s", fileTarget, err)
 	}
 
 	log.Printf("[DEBUG] Successfully deleted file %s", fileTarget)
 
 	return nil
 }
+
+func recursiveMkdir(d lxd.ContainerServer, container string, p string, mode os.FileMode, uid int64, gid int64) error {
```
Comment (sl1pm4t) [inline lxd/resource_lxd_container.go]: add comment that this func was lifted from LXD code

File: lxd/resource_lxd_container.go [hunk 11]
```diff
@@ -570,11 +646,60 @@ func resourceLxdContainerDeleteFile(client *lxd.Client, container string, file i
...
+
+	for ; i >= 1; i-- {
+		cur := filepath.Join(parts[:i]...)
+		_, resp, err := d.GetContainerFile(container, cur)
+		if err != nil {
+			continue
+		}
+
+		if resp.Type != "directory" {
+			return fmt.Errorf(i18n.G("%s is not a directory"), cur)
```
Comment (sl1pm4t) [inline lxd/resource_lxd_container.go]: remove il8n usage?

File: lxd/resource_lxd_volume_container_attach.go [hunk 12]
```diff
@@ -138,7 +145,7 @@ func resourceLxdVolumeContainerAttachDelete(d *schema.ResourceData, meta interfa
 		return err
 	}
 
-	v := NewVolumeAttachmentIdFromResourceId(d.Id())
+	// v := NewVolumeAttachmentIdFromResourceId(d.Id())
```
Comment (sl1pm4t) [inline lxd/resource_lxd_volume_container_attach.go]: tidy up

File: lxd/resource_lxd_container.go (hunk 5 above)
Comment (jtopjian) [inline lxd/resource_lxd_container.go]: Yeah, let's add the `refreshInterval` back here.

File: lxd/resource_lxd_container.go (hunk 11 above)
Comment (jtopjian) [inline lxd/resource_lxd_container.go]: I say yes to removing it - it's not used elsewhere in Terraform.

**Comment by jtopjian** [comment]
Nice work :)

**Comment by sl1pm4t** [comment]
Thanks for your help, figuring out the random failures.

**Comment by stgraber** [comment]
Awesome, thanks for porting!

**Comment by shantanugadgil** [comment]
@sl1pm4t 
While using this new plugin, can my Terraform binary continue to be 0.9.11?
Or, do I _have to_ upgrade by Terraform binary as well?

**Comment by jtopjian** [comment]
@shantanugadgil If you use the compiled binary that Matt just released, you have to upgrade to the Terraform 0.10 beta/rc.

If you wish to stay with your current version of Terraform, you will have to compile this from source, making sure you have tag `v0.9.11` checked out under `$GOPATH/src/github.com/hashicorp/terraform`.

**Comment by shantanugadgil** [comment]
Thanks @jtopjian .
On a hunch I just used the new 0.10.0beta2 release along with my TF 0.9.11 binary.
https://github.com/sl1pm4t/terraform-provider-lxd/releases/download/v0.10.0-beta2/terraform-provider-lxd_v0.10.0-beta2_linux_amd64.tar.gz

I got one error:
```
* provider.lxd: Could not read the lxc config: [/root/.config/lxc/config.yml]. Error: Unable to read the configuration file: open /root/.config/lxc/config.yml: no such file or directory
```
I touched that file to make the error go away.
Things don't seem awfully off for now.

But just to be safe, I'll update my TF binary as well.
I hope I'll get to test this soon, won't be able to test for a few days.

Regards,
Shantanu

**Pull request merged**